### PR TITLE
sdk: Extract hash and hasher crates

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Problem
The code in solana_program::hash is used all over the repo so it's one of the first things that needs to be made available outside of solana-program. 

#### Summary of Changes
- Add two new crates, `solana-hash` and `solana-hasher`. The `solana-hash` crate has no cryptography dependencies. The motivation is that sha2 takes a meaningful amount of time to compile and we use the `Hash` type in a lot of places where we don't compute the hash, e.g. parsing RPC responses.
- Move the `sol_sha256` syscall definition to `solana-hasher` to avoid depending on `solana-program`
- Re-export the public contents of `solana-hash` and `solana-hasher` (except for `sol_sha256`)  in `solana_program::hash` for backwards compatibility. I have not put a deprecation notice on the re-export because it's a core part of `solana-program`.
- Re-export `sol_sha256` in definitions.rs with a deprecation notice
- Make serde and bytemuck optional in `solana-hash`
- ~Make field 0 of the Hash struct pub (had no choice here)~ Update: I realised there was a way to keep field 0 `pub(crate)` when wasm_bindgen freaked out at the field being `pub`
- Make the MAX_BASE58_LEN const pub. I only needed to import it for tests but I think it's fine for this to be pub
- move the contents of `program/src/wasm/hash.rs` to the new `solana-hash` crate. This file was just an impl block so I don't think its removal requires any further re-exports
- replace solana-program dep in solana-merkle-tree with solana-hash and solana-hasher
- make the new crates no_std by default (the `solana-hash` crate gets a `std` feature)

### CODE REVIEW

**Comment by kevinheavey** [comment]
@ryoqun I'm quite stuck trying to understand this abi_digest test failure and was wondering if you could help:
```
thread 'blockhash_queue::BlockhashQueue_frozen_abi::test_abi_digest' panicked at accounts-db/src/blockhash_queue.rs:29:5:
assertion `left == right` failed: Possibly ABI changed? Confirm the diff by rerunning before and after this test failed with SOLANA_ABI_DUMP_DIR!
  left: "HzbCqb1YtCodkx6Fu57SpzkgaLp9WSrtw8texsjBhEDH"
 right: "BxykY65dC2NCcDm17rHQPjEY8wK55sKAhfhKVFGc5T1u"
```
This PR doesn't touch BlockhashQueue directly but it does move the `Hash` struct to its own crate. However I don't see how the Hash ABI could have changed? It's still just 
```
#[repr(transparent)]
pub struct Hash(pub(crate) [u8; HASH_BYTES]);
```
Is there more to it than that? Other types using Hash appear to also have changed digests

**Comment by kevinheavey** [comment]
OK the failing ABI tests passed when I temporarily added this to `AbiDigester::update` after the `normalize_type_name` call: 
```
.replace("solana_hash::Hash", "solana_program::hash::Hash")
```

So looks like the change in digest is just from moving the Hash struct and not from an actual ABI change

File: sdk/hash/Cargo.toml
Comment (buffalojoec) [inline sdk/hash/Cargo.toml]: We don't use a `bytemuck` feature for any other crates in the monorepo, and I'm not sure it's really necessary here. At a minimum, can we make it a default feature?

File: sdk/program/Cargo.toml
Comment (buffalojoec) [inline sdk/program/Cargo.toml]: I think we should sneak this crate's `borsh` dependency behind the program SDK's `borsh` feature so nobody is forced to bring in `borsh`.

**Review by buffalojoec — COMMENTED** [review]
Lgtm, just some small details, then we can ship it. Thanks for your patience on the few open PRs!

File: sdk/program/Cargo.toml
Comment (kevinheavey) [reply sdk/program/Cargo.toml]: good spot

File: sdk/hash/Cargo.toml
Comment (kevinheavey) [reply sdk/hash/Cargo.toml]: Ok so the compile times for small crates like this are dominated by the syn, quote and proc-macro2 crates, which bytemuck uses.

However thiserror also uses those crates, so I've now added a commit to replace thiserror with a cleaned-up version of what rust-analyzer expands the `derive(Error)` to. As you can see it's very simple.

I've also opened a [PR](https://github.com/anza-xyz/agave/pull/2165) to similarly remove thiserror from solana-sanitize, which solana-hash depends on. Once that's merged, solana-hash will not pull in syn and friends by default.

I would rather not make bytemuck a default feature because the impact on build times is quite substantial

**Comment by kevinheavey** [comment]
I think #2569 has been up for long enough to unpause this? @joncinque

File: merkle-tree/Cargo.toml
Comment (joncinque) [inline merkle-tree/Cargo.toml:14]: :partying_face:

File: sdk/hash/src/lib.rs
Comment (joncinque) [inline sdk/hash/src/lib.rs:1]: very slick! I didn't think it would be possible to make this `no_std`

File: sdk/hash/src/lib.rs
Comment (joncinque) [inline sdk/hash/src/lib.rs]: nit: for consistency with where this is needed at BorshSchema, we can do:
```suggestion
#[cfg(any(all(feature = "borsh", feature = "std"), target_arch = "wasm32"))]
use std::string::ToString;
```

Or conversely, simplify the other one to only require borsh -- whatever you prefer!

File: sdk/hash/src/lib.rs
Comment (joncinque) [inline sdk/hash/src/lib.rs]: To really be no-std, how about we do this with an array, ie:
```suggestion
        let mut bytes = [0; HASH_BYTES];
        let decoded_size = bs58::decode(s).onto(&mut bytes).map_err(|_| ParseHashError::Invalid)?;
        if decoded_size != mem::size_of::<Hash>() {
            Err(ParseHashError::WrongSize)
        } else {
            Ok(bytes.into())
        }
```

File: sdk/hash/src/lib.rs
Comment (joncinque) [inline sdk/hash/src/lib.rs:145]: This might cause too much churn, but this is really meant as a testing utility, and would be better served to live under a `dev-context-only-utils` feature. And that way `solana_atomic_u64` can also be optional. What do you think?

Alternatively, we can mark it as deprecated and say that people will need to use the feature soon?

File: sdk/hasher/Cargo.toml
Comment (joncinque) [inline sdk/hasher/Cargo.toml]: Since we'll likely also want to bust out keccak and blake3, how about making this more specific
```suggestion
name = "solana-sha256-hasher"
```

File: sdk/hasher/Cargo.toml
Comment (joncinque) [inline sdk/hasher/Cargo.toml]: Just so I'm sure I understand, is the sha2 feature and this dependency specifically for the case where someone might want to try to hash on-chain without the syscall?

I'm not too sure of a use-case for that, so we might want to prefer removing the option totally. If someone really wants to do it, they can do it themselves by pulling in sha2 directly in their program. This seems like more of a footgun than anything else.

Or is this to avoid breaking people who might be depending on this currently? If that's the case, how about adding a comment saying to remove this line during the next breaking release?

File: sdk/hasher/src/lib.rs
Comment (joncinque) [inline sdk/hasher/src/lib.rs]: love to see it

File: sdk/hasher/src/lib.rs
Comment (joncinque) [inline sdk/hasher/src/lib.rs]: Can you move these tests into `solana-hash`? Unless I've missed something, it looks like the only dependency on the hasher is the call to `hash(&[1u8])` at the start of `test_hash_fromstr`, so how about we remove that and just do a `Hash::new_from_array([1; 32])` for that initial hash.

File: sdk/hasher/Cargo.toml
Comment (joncinque) [inline sdk/hasher/Cargo.toml]: You'll be able to remove this by moving the tests in to solana-hash

**Review by joncinque — COMMENTED** [review]
This is really great work! Mostly small things that I hope will make it even better. Can you pass `solana-hash` to anza-team? Once we agree on a name for the hasher crate, we'll need to do the same

File: sdk/hasher/Cargo.toml
Comment (kevinheavey) [reply sdk/hasher/Cargo.toml]: > Or is this to avoid breaking people who might be depending on this currently? If that's the case, how about adding a comment saying to remove this line during the next breaking release?

This, there's no use case I can conceive of but it's a breaking change

File: sdk/hash/src/lib.rs
Comment (joncinque) [reply sdk/hash/src/lib.rs:145]: Ah never mind, sorry, this is breaking CI with:

> `dev-context-only-utils` must not be used as normal dependencies, but is by "([crate]: [dependency])":

We might just have to deprecate this separately

**Comment by kevinheavey** [comment]
I just noticed the Debug and Display impls were using the `std` feature of `bs58` so I changed them to avoid that

File: sdk/hash/src/lib.rs
Comment (joncinque) [inline sdk/hash/src/lib.rs]: nit: can we call this something with `write`, ie `write_as_base58` and move the parameters to look more like https://doc.rust-lang.org/std/fmt/fn.write.html? ie
```suggestion
fn write_as_base58(f: &mut fmt::Formatter, h: &Hash) -> fmt::Result {
```

File: sdk/hash/src/lib.rs
Comment (joncinque) [inline sdk/hash/src/lib.rs:73]: it's so strange that bs58 doesn't implement `onto` for a byte array, but oh well!

File: rpc-client-api/Cargo.toml
Comment (joncinque) [inline rpc-client-api/Cargo.toml:15]: It's definitely ok to have this, but just to check, this needs `std` because of std::error::Error, correct? https://github.com/Nullus157/bs58-rs/blob/e4a65a9ec641f26b61fb20518b51e83efd6352a0/src/lib.rs#L38

**Review by joncinque — COMMENTED** [review]
One last nit, then this should be good to go!

File: rpc-client-api/Cargo.toml
Comment (kevinheavey) [reply rpc-client-api/Cargo.toml:15]: correct

**Comment by kevinheavey** [comment]
@yihau I've sent invites to anza-team for the `solana-hash` and `solana-sha256-hasher` crates

File: sdk/hash/src/lib.rs
Comment (joncinque) [inline sdk/hash/src/lib.rs]: nit: sorry I only just noticed this, but the links don't make sense anymore, so best to remove the last bit of the sentence
```suggestion
/// This struct is used most often in `solana-sdk` and related crates to contain
/// a [SHA-256] hash, but may instead contain a [blake3] hash.
///
/// [SHA-256]: https://en.wikipedia.org/wiki/SHA-2
/// [blake3]: https://github.com/BLAKE3-team/BLAKE3
```

**Review by joncinque — COMMENTED** [review]
Looking really close! Sorry, I noticed a last thing on this pass

**Review by joncinque — COMMENTED** [review]
Ok, all good to go on my side! Once the crate check passes, I'll approve and we can merge

**Comment by yihau** [comment]
all pipelines are green now!

**Review by joncinque — APPROVED** [review]
Thanks for the work and addressing all the comments!

**Pull request merged**
